### PR TITLE
feat(subtitles): export translated srt

### PR DESCRIPTION
**Pull request by ringoshiina**
## Type of Changes

- [x] ✨ New feature (feat)
- [x] ✅ Test related (test)
- [x] 🌐 Internationalization (i18n)

## Description

Adds an on-demand "Download translated subtitles" action to the video subtitles settings panel. The export generates a complete AI-translated SRT from the full source subtitle track, not only cached playback translations.

This is a **smaller replacement** for #1603 (auto-closed by Contributor Trust at 1266 lines). This PR is **803 additions** across 17 files, well under the 1000-line threshold and below the 850-line target.

### What it does

- Translates all source subtitles via the existing Read Frog provider pipeline, then downloads as `.srt`
- **Fail-closed**: if any subtitle line fails to translate, the download is aborted, with no partial files
- **Same-language rejection**: if source and target languages match, the user is directed to "Download source subtitles" instead
- **Timing-gap fallback**: when AI segmentation creates coverage gaps, the export retries with smaller chunks; if gaps persist, it falls back to source-processed timing
- **Summary fetch tolerance**: a failed summary fetch does not block the export
- **Progress**: preparing → translating (N%) → complete, shown in the settings row
- **Error handling**: `toast.error` on failure, matching the existing source subtitle download pattern

### Feedback from #1603 addressed

| Codex P2 feedback on #1603 | Status in this PR |
| --- | --- |
| AI segmentation failure aborts entire export | ✅ Fixed: `buildExportProcessedChunk` catches errors and falls back to `optimizeSubtitles` |
| Incomplete batch wastes API quota translating remaining batches | ✅ Fixed: `hasIncompleteTranslatedFragments` is checked after each batch and throws immediately |
| AbortController/signal wiring issues | N/A: cancellation intentionally excluded (see below) |

### Small follow-up fixes in this PR

- Added a changeset for `@read-frog/extension`.
- Kept translated export summary fetching local to the export path so it cannot overwrite the live-translation summary guard state.

### Intentionally excluded (keeps PR small)

- Cancel button, `AbortController`/`AbortSignal`, navigation abort
- Preparation-phase progress details for segmentation and summary fetch
- Background/resumable export, persistent export state
- Filename suffix including target language code
- Toast error descriptions with actionable retry/provider guidance
- `CONTEXT.md` / documentation changes

The in-extension flow preserves Read Frog's subtitle normalization, provider configuration, and SRT timing instead of asking users to manually export source subtitles and translate them externally.

## Related Issue

Replaces #1603

## How Has This Been Tested?

- [x] Added unit tests
- [x] Verified through manual testing

**Automated**:

```
pnpm test src/entrypoints/subtitles.content/__tests__/universal-adapter.test.ts     # 12 passed
pnpm test .../__tests__/download-translated-subtitles.test.tsx                       # 2 passed
pnpm lint src/entrypoints/subtitles.content/universal-adapter.ts                     # passed
pnpm type-check                                                                      # passed
pre-push: lint, type-check, full test suite                                          # 148 files / 1258 tests passed
```

**Adapter test coverage**:

- Downloads complete translated SRT
- Rejects when any translated line is missing
- Rejects same-language export
- Reports progress after each batch
- Aborts before later batches when early batch is incomplete
- Falls back when AI segmentation throws
- Falls back when smaller AI chunks still have timing gaps
- Continues when summary fetch fails

**Manual testing** (`pnpm build` + `chrome-mv3`):

- Playback translation + translated SRT download confirmed working on YouTube

## Screenshots

N/A

## Checklist

- [x] I have tested these changes locally
- [x] I have updated the documentation accordingly if necessary
- [x] My code follows the code style of this project
- [x] My changes do not break existing functionality
- [x] If my code was generated by AI, I have proofread and improved it as necessary.

## Additional Information

**Line budget**: `17 files changed, 803 insertions(+), 6 deletions(-)` — targeting ≤850, hard limit <1000.

Potential follow-ups (separate PRs):

- Cancellation / AbortController wiring
- Preparation-phase progress feedback for segmentation and summary fetch
- Background/resumable export for very long videos
- Filename suffix including target language code
- Toast error descriptions with actionable next steps
- Direct unit tests for timing coverage helper boundary cases

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by taiiiyang** [comment]
我试了下在使用 ds flash 的情况下， 半个小时的视频需要2分多种才能下载完，我在想是不是直接把原字幕丢给 AI 会更快点

不过我更想了解下这个功能的使用场景主要是什么

如果可以的话，是否可以优化下下载的速度，比如说一次多翻译点之类的，我看了下沉浸式翻译其实也是有这个功能的

**Comment by ringoshiina** [comment]
@taiiiyang 感谢测试和反馈！

这个功能最初的使用场景是：我会用 `yt-dlp` 下载 YouTube 视频，但很多视频没有中文字幕，所以希望 Read Frog 能直接导出一个可以和下载视频放在一起使用的翻译 SRT。

我没有直接把 YouTube 原始字幕丢给 AI 翻译，主要是因为 YouTube 的字幕，尤其是自动字幕，经常更像流式字幕：片段很碎、相邻时间段有大量重复内容，直接翻译后即使能生成 SRT，外挂到视频上观感也比较差，常见问题是字幕边界不自然、重复句很多，甚至和画面/语音节奏不太匹配。

这个 PR 现在复用的是 Read Frog 已有的字幕处理流程：先通过现有 subtitle pipeline 做整理/分段，再走翻译和 SRT 导出。目标是生成一个更适合和下载视频一起使用的完整翻译字幕文件，而不是最快地把原始字幕文本翻译出来。

速度这点我认同还有优化空间。比如可以考虑增大单次翻译的字幕数量，或者后续支持并发 batch / 快速导出模式。为了避免这个 PR 再变大，我倾向于先把当前“可用且质量稳定”的导出流程合进去，然后把导出速度优化作为 follow-up PR 处理。

**Comment by taiiiyang** [comment]
> @taiiiyang 感谢测试和反馈！
> 
> 这个功能最初的使用场景是：我会用 `yt-dlp` 下载 YouTube 视频，但很多视频没有中文字幕，所以希望 Read Frog 能直接导出一个可以和下载视频放在一起使用的翻译 SRT。
> 
> 我没有直接把 YouTube 原始字幕丢给 AI 翻译，主要是因为 YouTube 的字幕，尤其是自动字幕，经常更像流式字幕：片段很碎、相邻时间段有大量重复内容，直接翻译后即使能生成 SRT，外挂到视频上观感也比较差，常见问题是字幕边界不自然、重复句很多，甚至和画面/语音节奏不太匹配。
> 
> 这个 PR 现在复用的是 Read Frog 已有的字幕处理流程：先通过现有 subtitle pipeline 做整理/分段，再走翻译和 SRT 导出。目标是生成一个更适合和下载视频一起使用的完整翻译字幕文件，而不是最快地把原始字幕文本翻译出来。
> 
> 速度这点我认同还有优化空间。比如可以考虑增大单次翻译的字幕数量，或者后续支持并发 batch / 快速导出模式。为了避免这个 PR 再变大，我倾向于先把当前“可用且质量稳定”的导出流程合进去，然后把导出速度优化作为 follow-up PR 处理。

感谢，我这两天来看下

File: src/entrypoints/subtitles.content/universal-adapter.ts
Comment (taiiiyang) [inline src/entrypoints/subtitles.content/universal-adapter.ts:1]: 我重新设计了一下，现在这一大堆逻辑不适合耦合在 adapter 中，我的建议还是拆出去做为一个 class，后续你做改动也方便，然后使用 atom 去存取相关的状态，遵循现在的项目里面的范式来重新设计，消除掉 clearSuccessTimeoutRef 之类杂乱的逻辑。
DownloadTranslatedSubtitles 这个组件尽量不要太多的逻辑，直接从 hook 里面取就好了，下面是相关的方案，可以参考下

1. **atom**(`atoms.ts`):phase 抽成 const 对象枚举 `TranslatedDownloadPhase`(`as const`,仓库惯例,非 TS enum)+ 同名类型,作为 class/atom/hook 共享来源;新增 `translatedSubtitlesDownloadStatusAtom`,存 `{ phase, progress: number|null }`。只存 phase,不存本地化字符串;三处都用常量,无裸字符串。
2. **新 class** `translated-subtitles-downloader.ts`:`constructor(fetcher: SubtitlesFetcher, config: PlatformConfig)`,`subtitlesStore` 直接 import 不注入; 私有字段 `isDownloading`、`successTimeout`(普通字段,非 ref);
3. 源字幕自理(`fetcher.fetch()` + 空值守卫 + `optimizeSubtitles`/preSegmented 处理);从 adapter 搬入导出逻辑(`buildCompleteTranslatedSubtitles` / `buildExportProcessedSubtitles` / `buildExportVideoContext` / 同语言守卫 / `TRANSLATION_BATCH_SIZE`);`download()` 防重入 → 推 phase/progress 到 atom → 成功 `complete`(4s 后回 idle)→ 失败 `toast.error` + `error`,`dispose()` 清 timeout;顺带修 review 两条隐患(config 开头快照一次、批次返回长度校验)。
4. **adapter**:`initialize()` 里 new 该 class;导航重置/销毁调 `dispose()`;`downloadTranslatedSubtitles` 委托给它;删搬走的方法。
5. **hook** `use-download-translated-subtitles.ts`:聚合展示逻辑,返回 `{ message, progress, isRunning, download }`。
6. **组件** `download-translated-subtitles.tsx`:删掉所有 state/effect/ref/disable,只解构 hook 纯渲染。

#### hook + 组件整合
```ts
// atoms.ts
export const TranslatedDownloadPhase = {
  Idle: "idle", Preparing: "preparing", Translating: "translating",
  Complete: "complete", Error: "error",
} as const
export type TranslatedDownloadPhase = typeof TranslatedDownloadPhase[keyof typeof TranslatedDownloadPhase]
export const translatedSubtitlesDownloadStatusAtom = atom<{
  phase: TranslatedDownloadPhase
  progress: number | null
}>({ phase: TranslatedDownloadPhase.Idle, progress: null })
```
```ts
// use-download-translated-subtitles.ts
export function useDownloadTranslatedSubtitles() {
  const status = useAtomValue(translatedSubtitlesDownloadStatusAtom)
  const { downloadTranslatedSubtitles } = useSubtitlesUI()
  const message = {
    [TranslatedDownloadPhase.Preparing]: i18n.t("subtitles.actions.downloadTranslatedPreparing"),
    [TranslatedDownloadPhase.Translating]: i18n.t("subtitles.actions.downloadTranslatedTranslating"),
    [TranslatedDownloadPhase.Complete]: i18n.t("subtitles.actions.downloadTranslatedComplete"),
    [TranslatedDownloadPhase.Idle]: null,
    [TranslatedDownloadPhase.Error]: null,
  }[status.phase]
  const isRunning = status.phase === TranslatedDownloadPhase.Preparing
    || status.phase === TranslatedDownloadPhase.Translating
  return { message, progress: status.progress, isRunning, download: downloadTranslatedSubtitles }
}
```
```tsx
// download-translated-subtitles.tsx —— 纯展示,无 state/effect/ref
export function DownloadTranslatedSubtitles() {
  const { message, progress, isRunning, download } = useDownloadTranslatedSubtitles()
  return (
    <SubtitlesSettingsItem
      icon={<IconLanguage className="size-4" />}
      label={(
        <div className="flex min-w-0 flex-col">
          <span className="truncate">{i18n.t("subtitles.actions.downloadTranslated")}</span>
          {message && (
            <span className="text-xs leading-4 text-yellow-300 drop-shadow-[0_1px_1px_rgba(0,0,0,0.85)]" aria-live="polite">
              {message}
              {progress !== null && ` (${progress}%)`}
            </span>
          )}
        </div>
      )}
      labelFor={buttonId}
    >
      <Button id={buttonId} type="button" variant="ghost-secondary" size="icon-sm" onClick={download} disabled={isRunning}>
        {isRunning ? <IconLoader2 className="size-3.5 animate-spin" /> : <IconDownload className="size-3.5" />}
      </Button>
    </SubtitlesSettingsItem>
  )
}
```

**Review by taiiiyang — COMMENTED** [review]
感谢贡献！
大致上没什么问题，不过架构上需要稍微调整下，确保不耦合在现在的 adapter 里面，然后也为了你后续调整方便一点。

File: src/entrypoints/subtitles.content/translated-subtitles-downloader.ts [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+      }
+    }
+  }
+
+  dispose(): void {
+    this.operationId++
+    this.isDownloading = false
+    this.clearSuccessTimeout()
+    this.setStatus(TranslatedDownloadPhase.Idle, null)
+  }
```
Comment (taiiiyang) [inline src/entrypoints/subtitles.content/translated-subtitles-downloader.ts:108]: dispose 这里需要调整下，现在 navigate 后，也就是 dispose，并不会去取消掉正在跑的请求，已经开始的下载流程还在继续跑。
可以想个办法去取消掉正在进行的请求，确保 navigate 后能够取消掉下载

**Comment by ringoshiina** [comment]
@taiiiyang 我同意这里应该继续改。

我看了一下现在的调用链：当前这版只是让 downloader 在 `dispose()` 后失效，避免旧任务继续更新状态或下载旧视频文件；但字幕翻译、summary、AI segmentation 实际都是通过 content script `sendMessage` 进入 background 的 `RequestQueue` / `BatchQueue`。所以如果要真正“取消正在跑的请求”，可能需要把 cancellation 往下打通到 background queue/message protocol，而不只是 downloader 层加判断。

这个改动范围可能会比现在大不少，也有可能让 PR 超过 Contributor Trust 里提到的 1000 changed lines 自动关闭阈值。为了避免我改完后 PR 被自动关掉，想先确认一下：

1. 这次是否可以先给这个 PR 加 `trust-check:skip`，或允许它为了 cancellation 超过 1000 行？
2. 如果希望继续保持小 PR，我也可以先只做 downloader 级别的 abort/失效，保证导航后不再继续推进后续 batch、不更新状态、不下载文件；更底层的 queue-level cancellation 后续单独开 PR。

你倾向哪种处理方式？

**Review by taiiiyang — APPROVED** [review]

**Comment by taiiiyang** [comment]
后续使用 codex 时，可以考虑删掉不太需要的 test 文件，不然不好 review

**Comment by ringoshiina** [comment]
@taiiiyang 感谢提醒，我后续会注意控制 Codex 生成的测试范围，尽量只保留关键行为测试，避免不必要的测试文件增加 review 成本。
